In [45]:
from pathlib import Path
import pandas as pd

data_path = Path("../data/raw")

datasets = {
    file.stem: pd.read_csv(file)
    for file in data_path.glob("*.csv")
}

In [46]:
l1 = list(datasets.keys())
print(l1)
i=0

['01_fund_master', '02_nav_history', '03_aum_by_fund_house', '04_monthly_sip_inflows', '05_category_inflows', '06_industry_folio_count', '07_scheme_performance', '08_investor_transactions', '09_portfolio_holdings', '10_benchmark_indices']


First CSV

In [47]:
df1 =datasets[l1[i]]
i=i+1

In [48]:
print(df1.shape)
print(df1.dtypes)


(40, 15)
amfi_code               int64
fund_house             object
scheme_name            object
category               object
sub_category           object
plan                   object
launch_date            object
benchmark              object
expense_ratio_pct     float64
exit_load_pct         float64
min_sip_amount          int64
min_lumpsum_amount      int64
fund_manager           object
risk_category          object
sebi_category_code     object
dtype: object


In [49]:
df1.head()

,amfi_code,fund_house,scheme_name,category,sub_category,plan,launch_date,benchmark,expense_ratio_pct,exit_load_pct,min_sip_amount,min_lumpsum_amount,fund_manager,risk_category,sebi_category_code
0,119551,SBI Mutual Fund,SBI Bluechip Fund - Regular Plan - Growth,Equity,Large Cap,Regular,2006-02-14,NIFTY 100 TRI,1.54,1.0,500,1000,Sohini Andani,Moderate,EC01
1,119552,SBI Mutual Fund,SBI Bluechip Fund - Direct Plan - Growth,Equity,Large Cap,Direct,2013-01-01,NIFTY 100 TRI,0.66,1.0,500,1000,Sohini Andani,Moderate,EC01
2,119598,SBI Mutual Fund,SBI Small Cap Fund - Regular Plan - Growth,Equity,Small Cap,Regular,2009-09-09,BSE 250 SmallCap TRI,1.43,1.0,500,1000,R. Srinivasan,Very High,EC03
3,119599,SBI Mutual Fund,SBI Small Cap Fund - Direct Plan - Growth,Equity,Small Cap,Direct,2013-01-01,BSE 250 SmallCap TRI,0.72,1.0,500,1000,R. Srinivasan,Very High,EC03
4,119120,SBI Mutual Fund,SBI Magnum Gilt Fund - Regular Plan - Growth,Debt,Gilt,Regular,2000-12-30,CRISIL Dynamic Gilt Index,0.77,0.0,500,1000,Dinesh Ahuja,Low,DC02


Second CSV

In [50]:
df2 = datasets[l1[i]]
i=i+1

In [51]:
print(df2.shape)
print(df2.dtypes)
print(df2.head())


(46000, 3)
amfi_code      int64
date          object
nav          float64
dtype: object
   amfi_code        date      nav
0     119551  2022-01-03  54.3856
1     119551  2022-01-04  54.3474
2     119551  2022-01-05  54.6869
3     119551  2022-01-06  55.4550
4     119551  2022-01-07  55.3692


Third CSV

In [52]:
df3= datasets[l1[i]]
i=i+1

In [53]:
print(df3.shape)
print(df3.dtypes)
print(df3.head())


(90, 5)
date               object
fund_house         object
aum_lakh_crore    float64
aum_crore           int64
num_schemes         int64
dtype: object
         date           fund_house  aum_lakh_crore  aum_crore  num_schemes
0  2022-03-31      SBI Mutual Fund            6.05     605000          186
1  2022-03-31  ICICI Prudential MF            4.65     465000          216
2  2022-03-31     HDFC Mutual Fund            4.35     435000          195
3  2022-03-31      Nippon India MF            2.70     270000          177
4  2022-03-31    Kotak Mahindra MF            2.70     270000          168


Fourth CSV

In [54]:
df4= datasets[l1[i]]
i=i+1

In [55]:
print(df4.shape)
print(df4.dtypes)
print(df4.head())


(48, 6)
month                         object
sip_inflow_crore               int64
active_sip_accounts_crore    float64
new_sip_accounts_lakh        float64
sip_aum_lakh_crore           float64
yoy_growth_pct               float64
dtype: object
     month  sip_inflow_crore  active_sip_accounts_crore  \
0  2022-01             11517                       4.91   
1  2022-02             11438                       4.93   
2  2022-03             12328                       5.09   
3  2022-04             11863                       5.48   
4  2022-05             12286                       5.55   

   new_sip_accounts_lakh  sip_aum_lakh_crore  yoy_growth_pct  
0                   9.10                4.80             NaN  
1                   8.20                4.85             NaN  
2                  10.50                5.01             NaN  
3                   9.52                5.12             NaN  
4                   8.10                5.15             NaN  


Fifth CSV

In [56]:
df5= datasets[l1[i]]
i=i+1

In [57]:
print(df5.shape)
print(df5.dtypes)
print(df5.head())


(144, 3)
month                object
category             object
net_inflow_crore    float64
dtype: object
     month         category  net_inflow_crore
0  2024-04        Large Cap            2413.0
1  2024-04          Mid Cap            3897.0
2  2024-04        Small Cap            3533.0
3  2024-04        Flexi Cap            4947.0
4  2024-04  Large & Mid Cap            4214.0


Sixth CSV

In [58]:
df6= datasets[l1[i]]
i=i+1

In [59]:
print(df6.shape)
print(df6.dtypes)
print(df6.head())


(21, 6)
month                   object
total_folios_crore     float64
equity_folios_crore    float64
debt_folios_crore      float64
hybrid_folios_crore    float64
others_folios_crore    float64
dtype: object
     month  total_folios_crore  equity_folios_crore  debt_folios_crore  \
0  2022-01               13.26                 9.28               1.86   
1  2022-04               13.91                 9.74               1.95   
2  2022-07               13.85                 9.69               1.94   
3  2022-10               14.12                 9.88               1.98   
4  2023-01               14.81                10.37               2.07   

   hybrid_folios_crore  others_folios_crore  
0                 0.80                 1.33  
1                 0.83                 1.39  
2                 0.83                 1.38  
3                 0.85                 1.41  
4                 0.89                 1.48  


Seventh CSV

In [60]:
df7= datasets[l1[i]]
i=i+1

In [61]:
print(df7.shape)
print(df7.dtypes)


(40, 19)
amfi_code               int64
scheme_name            object
fund_house             object
category               object
plan                   object
return_1yr_pct        float64
return_3yr_pct        float64
return_5yr_pct        float64
benchmark_3yr_pct     float64
alpha                 float64
beta                  float64
sharpe_ratio          float64
sortino_ratio         float64
std_dev_ann_pct       float64
max_drawdown_pct      float64
aum_crore               int64
expense_ratio_pct     float64
morningstar_rating      int64
risk_grade             object
dtype: object


In [62]:

df7.head()


,amfi_code,scheme_name,fund_house,category,plan,return_1yr_pct,return_3yr_pct,return_5yr_pct,benchmark_3yr_pct,alpha,beta,sharpe_ratio,sortino_ratio,std_dev_ann_pct,max_drawdown_pct,aum_crore,expense_ratio_pct,morningstar_rating,risk_grade
0,119551,SBI Bluechip Fund - Regular Plan - Growth,SBI Mutual Fund,Large Cap,Regular,12.42,12.36,14.45,11.49,0.87,0.89,0.88,1.29,14.0,-21.70,14288,1.54,4,Moderate
1,119552,SBI Bluechip Fund - Direct Plan - Growth,SBI Mutual Fund,Large Cap,Direct,15.25,11.30,14.23,9.52,1.78,0.87,0.81,1.29,14.0,-24.43,1231,0.66,3,Moderate
2,119598,SBI Small Cap Fund - Regular Plan - Growth,SBI Mutual Fund,Small Cap,Regular,24.56,23.39,20.67,22.16,1.23,0.89,0.94,1.35,25.0,-13.35,19259,1.43,5,Very High
3,119599,SBI Small Cap Fund - Direct Plan - Growth,SBI Mutual Fund,Small Cap,Direct,20.59,23.14,21.82,22.01,1.13,1.04,0.93,1.67,25.0,-24.78,36061,0.72,4,Very High
4,119120,SBI Magnum Gilt Fund - Regular Plan - Growth,SBI Mutual Fund,Gilt,Regular,5.34,6.07,5.43,4.47,1.60,0.22,1.52,2.11,4.0,-2.30,24101,0.77,5,Low


Eightth CSV

In [63]:
df8= datasets[l1[i]]
i=i+1

In [64]:
print(df8.shape)
print(df8.dtypes)



(32778, 13)
investor_id            object
transaction_date       object
amfi_code               int64
transaction_type       object
amount_inr              int64
state                  object
city                   object
city_tier              object
age_group              object
gender                 object
annual_income_lakh    float64
payment_mode           object
kyc_status             object
dtype: object


In [65]:
df8.head()

,investor_id,transaction_date,amfi_code,transaction_type,amount_inr,state,city,city_tier,age_group,gender,annual_income_lakh,payment_mode,kyc_status
0,INV003054,2024-01-01,119092,SIP,1834,Telangana,Hyderabad,T30,56+,Female,77.1,UPI,Verified
1,INV002952,2024-01-01,148567,Redemption,392882,Punjab,Amritsar,B30,18-25,Male,7.1,Cheque,Verified
2,INV003420,2024-01-01,118636,SIP,912,Haryana,Faridabad,B30,36-45,Male,47.2,Mandate,Verified
3,INV003436,2024-01-01,118634,SIP,1102,Maharashtra,Mumbai,T30,36-45,Female,54.4,Cheque,Pending
4,INV004691,2024-01-01,119094,Lumpsum,8682,Delhi,Noida,T30,26-35,Male,14.5,Net Banking,Pending


Ninth CSV

In [66]:
df9= datasets[l1[i]]
i=i+1

In [67]:
print(df9.shape)
print(df9.dtypes)
print(df9.head())


(322, 8)
amfi_code              int64
stock_symbol          object
stock_name            object
sector                object
weight_pct           float64
market_value_cr      float64
current_price_inr    float64
portfolio_date        object
dtype: object
   amfi_code stock_symbol                stock_name       sector  weight_pct  \
0     119551    POWERGRID    Power Grid Corporation    Utilities       13.85   
1     119551     HDFCBANK             HDFC Bank Ltd      Banking       11.19   
2     119551       GRASIM     Grasim Industries Ltd  Diversified        9.90   
3     119551      DRREDDY  Dr. Reddy's Laboratories       Pharma        4.76   
4     119551   ASIANPAINT          Asian Paints Ltd       Paints       10.25   

   market_value_cr  current_price_inr portfolio_date  
0           737.09            6011.08     2025-12-31  
1            88.97            1074.65     2025-12-31  
2           208.45            5964.59     2025-12-31  
3           161.32            3748.82     20

Tenth CSV

In [68]:
df10= datasets[l1[i]]
i=i+1

In [69]:
print(df10.shape)
print(df10.dtypes)
print(df10.head())


(8050, 3)
date            object
index_name      object
close_value    float64
dtype: object
         date index_name  close_value
0  2022-01-03    NIFTY50     17492.79
1  2022-01-04    NIFTY50     17689.64
2  2022-01-05    NIFTY50     17835.05
3  2022-01-06    NIFTY50     17878.51
4  2022-01-07    NIFTY50     17759.15


Anamoly Detection

In [86]:
for name, df in datasets.items():
    print(f"\n{name}")
    print(df.isnull().sum())


01_fund_master
amfi_code             0
fund_house            0
scheme_name           0
category              0
sub_category          0
plan                  0
launch_date           0
benchmark             0
expense_ratio_pct     0
exit_load_pct         0
min_sip_amount        0
min_lumpsum_amount    0
fund_manager          0
risk_category         0
sebi_category_code    0
dtype: int64

02_nav_history
amfi_code    0
date         0
nav          0
dtype: int64

03_aum_by_fund_house
date              0
fund_house        0
aum_lakh_crore    0
aum_crore         0
num_schemes       0
dtype: int64

04_monthly_sip_inflows
month                         0
sip_inflow_crore              0
active_sip_accounts_crore     0
new_sip_accounts_lakh         0
sip_aum_lakh_crore            0
yoy_growth_pct               12
dtype: int64

05_category_inflows
month               0
category            0
net_inflow_crore    0
dtype: int64

06_industry_folio_count
month                  0
total_folios_crore     

In [87]:
for name, df in datasets.items():
    print(f"{name}: {df.duplicated().sum()} duplicate rows")

01_fund_master: 0 duplicate rows
02_nav_history: 0 duplicate rows
03_aum_by_fund_house: 0 duplicate rows
04_monthly_sip_inflows: 0 duplicate rows
05_category_inflows: 0 duplicate rows
06_industry_folio_count: 0 duplicate rows
07_scheme_performance: 0 duplicate rows
08_investor_transactions: 0 duplicate rows
09_portfolio_holdings: 0 duplicate rows
10_benchmark_indices: 0 duplicate rows


## Observed Anomalies

1. Date columns are stored as object datatype.
2. yoy_growth_pct contains missing values.
3. No duplicate AMFI codes detected.
4. All datasets loaded successfully.
5. AMFI code coverage is 100%.

Analysis

In [70]:
fund_master = datasets['01_fund_master']
nav_history = datasets['02_nav_history']

Fund Master Exploration

In [73]:
print("="*50)
print("FUND MASTER EXPLORATION")
print("="*50)

print("\nUnique Fund Houses:")
print(fund_master['fund_house'].nunique())
print(fund_master['fund_house'].unique())

print("\nUnique Categories:")
print(fund_master['category'].nunique())
print(fund_master['category'].unique())

print("\nUnique Sub-Categories:")
print(fund_master['sub_category'].nunique())
print(fund_master['sub_category'].unique())

print("\nRisk Grade Distribution:")
print(fund_master['risk_category'].value_counts())

FUND MASTER EXPLORATION

Unique Fund Houses:
10
['SBI Mutual Fund' 'HDFC Mutual Fund' 'ICICI Prudential MF'
 'Nippon India MF' 'Kotak Mahindra MF' 'Axis Mutual Fund'
 'Aditya Birla Sun Life MF' 'UTI Mutual Fund' 'Mirae Asset MF'
 'DSP Mutual Fund']

Unique Categories:
2
['Equity' 'Debt']

Unique Sub-Categories:
12
['Large Cap' 'Small Cap' 'Gilt' 'Mid Cap' 'Short Duration' 'Value'
 'Liquid' 'Index/ETF' 'Flexi Cap' 'Index' 'Large & Mid Cap' 'ELSS']

Risk Grade Distribution:
risk_category
Moderate           16
High                8
Very High           6
Low                 6
Moderately High     4
Name: count, dtype: int64


Understand AMFI Scheme Codes

In [74]:
print("="*50)
print("AMFI SCHEME CODE ANALYSIS")
print("="*50)

print("\nSample AMFI Codes:")

print(
    fund_master[
        ['amfi_code','scheme_name']
    ].head(10)
)

print("\nTotal Unique AMFI Codes:")
print(fund_master['amfi_code'].nunique())

AMFI SCHEME CODE ANALYSIS

Sample AMFI Codes:
   amfi_code                                        scheme_name
0     119551          SBI Bluechip Fund - Regular Plan - Growth
1     119552           SBI Bluechip Fund - Direct Plan - Growth
2     119598         SBI Small Cap Fund - Regular Plan - Growth
3     119599          SBI Small Cap Fund - Direct Plan - Growth
4     119120       SBI Magnum Gilt Fund - Regular Plan - Growth
5     100016          HDFC Top 100 Fund - Regular Plan - Growth
6     125497           HDFC Top 100 Fund - Direct Plan - Growth
7     100033  HDFC Mid-Cap Opportunities Fund - Regular - Gr...
8     125498  HDFC Mid-Cap Opportunities Fund - Direct - Growth
9     100025       HDFC Short Term Debt Fund - Regular - Growth

Total Unique AMFI Codes:
40


### AMFI Scheme Code Structure

AMFI (Association of Mutual Funds in India) assigns a unique numeric
scheme code to each mutual fund scheme.

The AMFI code serves as:
- Unique scheme identifier
- Primary key for linking datasets
- Reference used by NAV providers and APIs
- Join key between fund master and NAV history

Example:
125497 → HDFC Top 100 Fund Direct Growth
119551 → SBI Bluechip Fund

Validate AMFI Codes

In [75]:
fund_codes = set(
    fund_master['amfi_code']
)

nav_codes = set(
    nav_history['amfi_code']
)

matching_codes = fund_codes.intersection(nav_codes)

missing_codes = fund_codes - nav_codes

extra_codes = nav_codes - fund_codes

Validation Summary

In [76]:
print("="*50)
print("AMFI CODE VALIDATION")
print("="*50)

print(f"Fund Master Codes : {len(fund_codes)}")
print(f"NAV History Codes : {len(nav_codes)}")
print(f"Matching Codes    : {len(matching_codes)}")
print(f"Missing Codes     : {len(missing_codes)}")
print(f"Extra Codes       : {len(extra_codes)}")

coverage = (
    len(matching_codes)
    / len(fund_codes)
) * 100

print(f"\nCoverage: {coverage:.2f}%")

AMFI CODE VALIDATION
Fund Master Codes : 40
NAV History Codes : 40
Matching Codes    : 40
Missing Codes     : 0
Extra Codes       : 0

Coverage: 100.00%


Display Missing Codes (if any)

In [77]:
if len(missing_codes) > 0:
    print("\nMissing Codes:")
    print(list(missing_codes)[:20])
else:
    print("\nAll AMFI codes are present in NAV history.")


All AMFI codes are present in NAV history.


## Data Quality Summary

### Findings

- Total schemes in fund_master:40
- Total schemes in nav_history:40
- Matching AMFI codes: 40
- Coverage: 100.00%

### Observations

- AMFI codes are unique identifiers for mutual fund schemes.
- Fund houses, categories, sub-categories and Risk Categories were successfully explored.
- AMFI codes were validated between fund_master and nav_history datasets.
- No major structural issues were observed.
- AMFI code can be safely used as the primary key for joining mutual fund datasets.

### Conclusion

The datasets were successfully loaded and explored. The AMFI scheme code provides a reliable linkage between master fund information and NAV history data, making it suitable for downstream analytics and reporting.